crates/config-encryption/src/encrypt/mod.rs

```diff
@@ -0,0 +1,214 @@
+            "--input-type=json",
+            "--output-type",
+            output_format.sops_type(),
+            "/dev/stdin",
+        ])
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    serde_json::to_writer(child.stdin.take().unwrap(), &plain_config.plain_json)?;
+    drop_plain_config(plain_config.plain_json);
+
+    let Output {
+        status,
+        stdout,
+        stderr,
+    } = child.wait_with_output()?;
+
+    if status.success() {
+        Ok(EncryptedConfig {
+            document: stdout,
+            format: output_format,
+        })
+    } else {
+        let sops_stderr = String::from_utf8_lossy(&stderr).to_string();
+        Err(Error::SopsFailed(status, sops_stderr))
+    }
+}
+
+fn add_suffix_to_location(
+    conf: &mut serde_json::Value,
+    location_ptr: &str,
+    suffix: &str,
+) -> Result<(), Error> {
+    let mut parent = conf;
+    let parsed_ptr = Pointer::from_str(location_ptr);
+    let mut iter = parsed_ptr.iter().peekable();
+    let mut leaf_prop = None;
+    while let Some(token) = iter.next() {
+        if iter.peek().is_none() {
+            leaf_prop = Some(token);
+            break;
+        }
+        match token {
+            Token::Property(p) => parent = parent.get_mut(p).expect("deref property"),
+            Token::Index(i) => parent = parent.get_mut(i).expect("deref index"),
+            Token::NextIndex => panic!("nextIndex is invalid"),
+        }
+    }
+
+    // parent must be an object, since there's no way to add a suffix to array indices. If it's
+    // not, then we return an error.
+    let parent_obj = parent.as_object_mut().ok_or_else(|| {
+        Error::InvalidSecretLocation(location_ptr.to_owned(), "it is not an object property")
+    })?;
+
+    // If ptr is an empty string, then parent will point to the root document, and leaf_prop will
+    // be None. If ptr is non-empty, then leaf_prop must be Some.
+    if let Some(token) = leaf_prop {
+        if let Token::Property(prop) = token {
+            let (mut key, value) = parent_obj.remove_entry(prop).unwrap();
+            key.push_str(suffix);
+            parent_obj.insert(key, value);
+        } else {
+            panic!(
+                "expected leaf token to be an object property, but was {:?}",
+                token
+            );
+        }
+    } else {
+        return Err(Error::InvalidSecretLocation(
+            location_ptr.to_owned(),
+            "it is the root document",
+        ));
+    }
+    Ok(())
+}
+
+/// We'd like to avoid keeping sensitive values around in memory, but `serde_json::Value` doesn't
+/// have an easy way to zero out the whole document, so this just traverses the object and
+/// overwrites all the String values. There's no safe way to zero numbers, but they seem far less
+/// likely to hold sensitive information anyway.
+fn drop_plain_config(json: serde_json::Value) {
```

Comment:
Not sure if this helps or not, but you might be able to use https://crates.io/crates/secrecy to help here. It supports deserializing into a `Secret` wrapper that can help with accidentally mishandling a sensitive value. I haven't used it, but it's something you might find interesting.

Reply:
Yeah I had hoped to be able to use something like that, but the reason it doesn't work is that it only handles owned contiguous memory regions. So it won't actually zero the memory of all the nested values when it's dropped :( I could see it still being useful just as a way of communicating to developers that it's sensitive, though. Thanks :+1: